prometheus/src/main/java/org/opensearch/sql/prometheus/storage/implementor/PrometheusDefaultImplementor.java

```diff
@@ -31,26 +43,104 @@
+    String step = context.getRequest().getStep();
+    String seriesSelectionQuery
+        = SeriesSelectionQueryBuilder.build(node.getMetricName(), node.getFilter());
+
+    String aggregateQuery
+        = AggregationQueryBuilder.build(node.getAggregatorList(),
+        node.getGroupByList());
+
+    String finalQuery = String.format(aggregateQuery, seriesSelectionQuery + "[" + step + "]");
+    context.getRequest().setPromQl(finalQuery);
+
+    //Since prometheus response doesn't have any fieldNames in its output.
+    //the field names are sent to PrometheusResponse constructor via context.
+    setPrometheusResponseFieldNames(node, context);
+    return context;
+  }
+
   @Override
   public PhysicalPlan visitRelation(LogicalRelation node,
                                     PrometheusMetricScan context) {
+    PrometheusMetricTable prometheusMetricTable = (PrometheusMetricTable) node.getTable();
+    if (prometheusMetricTable.getMetricName() != null) {
+      String query = SeriesSelectionQueryBuilder.build(node.getRelationName(), null);
+      context.getRequest().setPromQl(query);
+      setTimeRangeParameters(null, context);
+      context.getRequest()
+          .setStep(StepParameterResolver.resolve(context.getRequest().getStartTime(),
+              context.getRequest().getEndTime(), null));
+    }
     return context;
   }
 
-  // Since getFieldTypes include labels
-  // we are explicitly specifying the output column names;
-  @Override
-  public PhysicalPlan visitProject(LogicalProject node, PrometheusMetricScan context) {
-    List<NamedExpression> finalProjectList = new ArrayList<>();
-    finalProjectList.add(
-        new NamedExpression(LABELS, new ReferenceExpression(LABELS, ExprCoreType.STRING)));
-    finalProjectList.add(
-        new NamedExpression(TIMESTAMP,
-            new ReferenceExpression(TIMESTAMP, ExprCoreType.TIMESTAMP)));
-    finalProjectList.add(
-        new NamedExpression(VALUE, new ReferenceExpression(VALUE, ExprCoreType.DOUBLE)));
-    return new ProjectOperator(visitChild(node, context), finalProjectList,
-        node.getNamedParseExpressions());
+  private void setTimeRangeParameters(Expression filter, PrometheusMetricScan context) {
+    TimeRangeParametersResolver timeRangeParametersResolver = new TimeRangeParametersResolver();
+    Pair<Long, Long> timeRange = timeRangeParametersResolver.resolve(filter);
+    context.getRequest().setStartTime(timeRange.getFirst());
+    context.getRequest().setEndTime(timeRange.getSecond());
   }
 
+  private void setPrometheusResponseFieldNames(PrometheusLogicalMetricAgg node,
+                                               PrometheusMetricScan context) {
+    Optional<NamedExpression> spanExpression = getSpanExpression(node.getGroupByList());
+    if (spanExpression.isEmpty()) {
+      throw new RuntimeException(
+          "Prometheus Catalog doesn't support aggregations without span expression");
+    }
```

Comment:
If span is required, the exception should be thrown at logical -> physical stage, instead of here.

Reply:
Right, but that would cause issues with opensearch catalog,,,since it would be allowed for them. Don't want to distinguish catalogs in Analyzer component.

Can I do somewhere else?

Introduce validation of logical plan?